bdai_ros2_wrappers/bdai_ros2_wrappers/action.py

```diff
@@ -0,0 +1,583 @@
+
+    def write(self, data: Any) -> None:
+        """Write the data tape."""
+        with self._lock:
+            for stream in self._streams:
+                if not stream.write(data):
+                    message = "Stream is filled up, dropping message"
+                    if stream.label:
+                        message = f"{stream.label}: {message}"
+                    warnings.warn(message, RuntimeWarning)
+            self._content.append(data)
+
+    def content(
+        self,
+        *,
+        follow: bool = False,
+        follow_buffer_size: Optional[int] = None,
+        timeout_sec: Optional[float] = None,
+        label: Optional[str] = None,
+    ) -> Iterable:
+        """
+        Iterate over the data tape.
+
+        When following the data tape, iteration stops when the given timeout
+        expires and when the data tape is closed.
+
+        Args:
+            follow: whether to follow the data tape as it gets written or not.
+            follow_buffer_size: optional buffer size when following the data tape.
+            If none is provided, the data tape size will be used as default.
+            timeout_sec: optional timeout, in seconds, when following the data tape.
+            label: optional label to qualify logs and warnings.
+
+        Returns:
+            a lazy iterable over the data tape.
+        """
+        with self._lock:
+            content = self._content.copy()
+            stream: Optional[Tape.Stream] = None
+            if follow and not self._closed:
+                stream = Tape.Stream(follow_buffer_size, label)
+                self._streams.append(stream)
+        try:
+            yield from content
+            if stream is not None:
+                while not self._closed:
+                    feedback = stream.read(timeout_sec)
+                    if feedback is None:
+                        break
+                    yield feedback
+                while not stream.consumed:
+                    # This is safe as long as there is
+                    # a single reader for the stream,
+                    # which is currently the case.
+                    feedback = stream.read(timeout_sec)
+                    if feedback is None:
+                        continue
+                    yield feedback
+        finally:
+            if stream is not None:
+                with self._lock:
+                    self._streams.remove(stream)
+
+    def close(self) -> None:
+        """
+        Close the data tape.
+
+        This will interrupt all following content iterators.
+        """
+        self._closed = True
+        with self._lock:
+            for stream in self._streams:
+                stream.interrupt()
+
+
+class ActionFuture:
+    """
+    A proxy to a ROS 2 action invocation.
+
+    Action futures are to actions what plain futures are to services, with a bit more functionality
+    to cover action specific semantics such feedback and cancellation.
```

Comment:
```suggestion
    to cover action specific semantics such as feedback and cancellation.
```

Reply:
Fixed in https://github.com/bdaiinstitute/ros_utilities/pull/58/commits/329efc3935f1f9fdc187d6157e4d218f4289adfd.